query-parser/src/parse.rs

```diff
@@ -110,21 +113,104 @@ def_value_satisfy_parser_fn!(Where,
                              PatternNonValuePlace,
                              PatternNonValuePlace::from_value);
 
+fn seq<T: Into<Option<edn::Value>>>(x: T) -> Option<Vec<edn::Value>> {
+    match x.into() {
+        Some(edn::Value::List(items)) => Some(items.into_iter().collect()),
+        Some(edn::Value::Vector(items)) => Some(items),
+        _ => None,
+    }
+}
+
 /// Take a vector Value containing one vector Value, and return the `Vec` inside the inner vector.
 /// Also accepts an inner list, returning it as a `Vec`.
 fn unwrap_nested(x: edn::Value) -> Option<Vec<edn::Value>> {
     match x {
         edn::Value::Vector(mut v) => {
-            match v.pop() {
-                Some(edn::Value::List(items)) => Some(items.into_iter().collect()),
-                Some(edn::Value::Vector(items)) => Some(items),
-                _ => None,
-            }
+            seq(v.pop())
         }
         _ => None,
     }
 }
 
+def_value_parser_fn!(Where, and, (), input, {
+    matches_plain_symbol!("and", input)
+});
+
+def_value_parser_fn!(Where, or, (), input, {
+    matches_plain_symbol!("or", input)
+});
+
+def_value_parser_fn!(Where, or_join, (), input, {
+    matches_plain_symbol!("or-join", input)
+});
+
+def_value_parser_fn!(Where, rule_vars, Vec<Variable>, input, {
+    satisfy_map(|x: edn::Value| {
```

Comment:
It took a long time, but I think I've figured out improvements to this pattern, which should allow us to write something like:
```rust
or(vector(), list()).of(many1::<Vec<_>, _>(Query::variable())
```
which I think is a significant improvement.